src/main/java/com/microsoft/sqlserver/jdbc/KerbAuthentication.java

```diff
@@ -300,8 +308,121 @@ private String makeSpn(String server, int port) throws SQLServerException
+                // Since we have a match, our hostname is the correct one (for instance of server
+                // name was an IP), so we override dnsName as well
+                dnsName = canonicalHostName;
+            } catch (UnknownHostException cannotCanonicalize) {
+                // ignored, but we are in a bad shape
+            }
+        }
+        if (realm == null) {
+            return spn;
+        } else {
+            StringBuilder sb = new StringBuilder("MSSQLSvc/");
+            sb.append(dnsName).append(":").append(portOrInstance).append("@").append(realm.toUpperCase(Locale.ENGLISH));
+            return sb.toString();
+        }
+    }
+
+    private static RealmValidator validator;
+
+    /**
+     * Find a suitable way of validating a REALM for given JVM.
+     *
+     * @return a not null realm Validator.
+     */
+    static RealmValidator getRealmValidator() {
+        if (validator != null) {
+            return validator;
+        }
+        // JVM Specific, here Sun/Oracle JVM
+        try {
+            Class<?> clz = Class.forName("sun.security.krb5.Config");
+            Method getInstance = clz.getMethod("getInstance", new Class[0]);
+            final Method getKDCList = clz.getMethod("getKDCList", new Class[] { String.class });
+            final Object instance = getInstance.invoke(null);
+            RealmValidator oracleRealmValidator = new RealmValidator() {
+
+                @Override
+                public boolean isRealmValid(String realm) {
+                    try {
+                        Object ret = getKDCList.invoke(instance, realm);
+                        return ret != null;
+                    } catch (Exception err) {
+                        return false;
+                    }
+                }
+            };
+            validator = oracleRealmValidator;
+            return oracleRealmValidator;
+        } catch (ReflectiveOperationException notTheRightJVMException) {
+            // Ignored, we simply are not using the right JVM
+        }
+        // No implementation found, default one, not any realm is valid
+        validator = new RealmValidator() {
+            @Override
+            public boolean isRealmValid(String realm) {
+                return false;
+            }
+        };
+        return validator;
+    }
+
+    /**
+     * Try to find a REALM in the different parts of a host name.
+     *
+     * @param realmValidator a function that return true if REALM is valid and exists
+     * @param hostname the name we are looking a REALM for
+     * @return the realm if found, null otherwise
+     */
+    private static String findRealmFromHostname(RealmValidator realmValidator, String hostname) {
+        if (hostname == null) {
+            return null;
+        }
+        int index = 0;
+        while (index != -1 && index < hostname.length() - 2) {
+            String realm = hostname.substring(index + 1);
+            if (realmValidator.isRealmValid(realm)) {
```

Comment:
Ok, tested on my side, but it does not change anything. The result is still the same, it works.

However, I don't understand your patch :

In realm validator, we are supposed to validate a sub part of the hostname until we find a valid realm, but you propose to test against the default realm each time after testing against the realm we are currently testing ?

This test could be DONE outside the realm validator and if successful, simply return a always false validator.

In order to investigate this more closely, I have this test example:
```java
SQLServerDataSource msSQLdataSource = new SQLServerDataSource();    
msSQLdataSource.setURL("jdbc:sqlserver://"+ipOrHostname+":1033;Database=MyDatabase;integratedSecurity=true;authenticationScheme=JavaKerberos");
msSQLdataSource.getConnection().isValid(5);
System.out.println("is ok")
```

We have 2 DC with realms NY8.AD.MYCOMPANY.COM and PAR.AD.MYCOMPANY.COM

**Test with IP address**
The result (with a few debug messages) is the following with ipOrHostname = 10.50.4.6
```
realmValidator: Checking realm=0.50.4.6	; default_realm=NY8.AD.MYCOMPANY.COM
realmValidator: Checking realm=50.4.6	; default_realm=NY8.AD.MYCOMPANY.COM
realmValidator: Checking realm=4.6	; default_realm=NY8.AD.MYCOMPANY.COM
REVERSE DNS FOUND: sql07.par.ad.mycompany.com
realmValidator: Checking realm=sql07.par.ad.mycompany.com	; default_realm=NY8.AD.MYCOMPANY.COM
realmValidator: Checking realm=par.ad.mycompany.com	; default_realm=NY8.AD.MYCOMPANY.COM
SPN before enrichment: MSSQLSvc/10.50.4.6:1033 ; AFTER enrichment: MSSQLSvc/sql07.par.ad.mycompany.com:1033@PAR.AD.MYCOMPANY.COM
is ok
```

I saw incidentally during this test that we should not add +1 at line 410 when index is 0... it may cause issue if the SQL server is on the Active Directory machine (Is it you case btw?)

**Test with FQDN**
The same example with the FQDN name of machine (not its IP address): sql07.par.ad.mycompany.com

```
Checking realm=sql07.par.ad.mycompany.com	; default_realm=NY8.AD.MYCOMPANY.COM
Checking realm=par.ad.mycompany.com	; default_realm=NY8.AD.MYCOMPANY.COM
SPN before enrichment: MSSQLSvc/sql07.par.ad.mycompany.com:1033 ; AFTER enrichment: MSSQLSvc/sql07.par.ad.mycompany.com:1033@PAR.AD.MYCOMPANY.COM
```

**Test with FQDN in same realm**

Same 2 examples, but I changed the default realm and my identity to use the same realm as the MS SQL server (and an identity on the same realm):
```
Checking realm=0.50.4.6	; default_realm=PAR.AD.MYCOMPANY.COM
Checking realm=50.4.6	; default_realm=PAR.AD.MYCOMPANY.COM
Checking realm=4.6	; default_realm=PAR.AD.MYCOMPANY.COM
Checking realm=sql07-par.par.ad.mycompany.com	; default_realm=PAR.AD.MYCOMPANY.COM
Checking realm=par.ad.mycompany.com	; default_realm=PAR.AD.MYCOMPANY.COM
SPN before enrichment: MSSQLSvc/10.50.4.6:1033 ; AFTER enrichment: MSSQLSvc/sql07.par.ad.mycompany.com:1033@PAR.AD.MYCOMPANY.COM
is ok
Checking realm=ql07-par.mycompany.com	; default_realm=PAR.AD.MYCOMPANY.COM
Checking realm=par.ad.mycompany.com	; default_realm=PAR.AD.MYCOMPANY.COM
SPN before enrichment: MSSQLSvc/sql07.par.ad.mycompany.com:1033 ; AFTER enrichment: MSSQLSvc/sql07.par.ad.mycompany.com:1033@PAR.AD.MYCOMPANY.COM
is ok
```

**Possible explanation of the behaviour you have in your lab**

I don't understand the result. Factually, there is a bug in my implementation in the case where the SQL Server is on the same machine as the realm since it will cut the first char of the domain name. This bug may trigger something tricky if the domain of the SQL Server you are trying to reach is part of a wider realm however.

I imagine you have a domain such as :

DEV.MSFT.COM

and you create another realm with trust called:

MYLAB.DEV.MSFT.COM

and you install an SQL server on it and use mylab.dev.msft.com as server name.

In that case, at in the realm validator, it won't check mylab.dev.msft.com but will check ylab.dev.microsoft.com instead -> no domain found and then will find dev.msft.com to be the correct domain -> realm DEV.MSFT.COM is correct -> it would explain the bug.

Am I correct ?

In that case, the fix is trivial, at line 410, replace
`String realm = hostname.substring(index + 1);`
by
`String realm = hostname.substring(index);`
and at line 414, replace:
`index = hostname.indexOf(".", index + 1);`
by
```
index = hostname.indexOf(".", index + 1);
if (index != -1)
 index++;
```

If it solves your issue, that's cool, in other case, I still don't understand the issue and I don't get your proposal for patch.

I will soon update the MR with my change.

Regards